index.js

```diff
@@ -1,262 +1,515 @@
+ * `MongoClient.connect(model.driver_url, model.driver_options)`.
+ */
+assign(derived, {
+  /**
+   * Get the URL which can be passed to `MongoClient.connect`.
+   * @see http://bit.ly/mongoclient-connect
+   * @return {String}
+   */
+  driver_url: {
+    deps: [
+      'hostname',
+      'port',
+      'ssl',
+      'kerberos_principal',
+      'kerberos_password',
+      'kerberos_service_name',
+      'mongodb_username',
+      'mongodb_password',
+      'mongodb_database_name',
+      'driver_auth_mechanism'
+    ],
+    fn: function() {
+      var req = {
+        protocol: 'mongodb',
+        slashes: true,
+        hostname: this.hostname,
+        port: this.port,
+        query: {
+          slaveOk: 'true'
+        }
+      };
+
+      if (this.ssl) {
+        req.query.ssl = 'true';
+      }
+
+      if (this.authentication === 'NONE') {
+        return toURL(req);
       }
-    }
 
-    if (attrs.gssapi_service_name) {
-      if (attrs.auth_mechanism !== 'GSSAPI') {
-        return new TypeError(format(
-          'The `gssapi_service_name` field does not apply when '
-          + 'using `%s` as the auth mechanism.', attrs.auth_mechanism));
+      if (this.authentication === 'MONGODB') {
+        assign(req, {
+          auth: format('%s:%s', this.mongodb_username, this.mongodb_password),
+          query: {
+            slaveOk: 'true',
+            authSource: this.mongodb_database_name
+          }
+        });
+        return toURL(req);
       }
+
+      if (this.authentication === 'KERBEROS') {
+        assign(req, {
+          pathname: 'kerberos',
+          query: {
+            slaveOk: 'true',
+            gssapiServiceName: this.kerberos_service_name,
+            authMechanism: this.driver_auth_mechanism
+          }
+        });
+
+        if (this.kerberos_password) {
+          req.auth = format('%s:%s',
+            encodeURIComponent(this.kerberos_principal),
+            this.kerberos_password);
+        } else {
+          req.auth = format('%s',
+            encodeURIComponent(this.kerberos_principal));
+        }
+        return toURL(req);
+      }
+
+      throw new TypeError('Unspported authentication method.');
     }
+  },
+  /**
+   * Get the options which can be passed to `MongoClient.connect`
+   * in addition to the URI.
+   * @see http://mongodb.github.io/node-mongodb-native/2.0/api/MongoClient.html#.connect
+   * @return {Object}
+   */
+  driver_options: {
+    deps: [
+      'ssl_validate',
+      'ssl_ca',
+      'ssl_cert',
+      'ssl_private_key',
+      'ssl_private_key_password'
+    ],
+    fn: function() {
+      var connectionOptions = {
+        uri_decode_auth: true,
+        db: {},
+        server: {},
+        replSet: {
+          connectWithNoPrimary: true
+        },
+        mongos: {}
+      };
 
-    if (!attrs.ssl) {
-      if (attrs.ssl_validate) {
-        return new TypeError('The `ssl_validate` field requires `ssl = true`.');
```

Comment:
Another thing we could do is put all of this in a validation section. We would have to check if it's valid, but we actually may be able to do that from this function also.